internal/cri/server/container_image_mount.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/containerd/containerd/v2/defaults"
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+	"github.com/containerd/platforms"
+	"github.com/opencontainers/image-spec/identity"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+const (
+	labelGCSnapRef = "containerd.io/gc.ref.snapshot."
+)
+
+func (c *criService) mutateMounts(
+	ctx context.Context,
+	extraMounts []*runtime.Mount,
+	snapshotter string,
+	sandboxID string,
+	platform imagespec.Platform,
+) error {
+	for _, m := range extraMounts {
+		err := c.mutateImageMount(ctx, m, snapshotter, sandboxID, platform)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *criService) mutateImageMount(
+	ctx context.Context,
+	extraMount *runtime.Mount,
+	snapshotter string,
+	sandboxID string,
+	platform imagespec.Platform,
+) (retErr error) {
+	imageSpec := extraMount.GetImage()
+	if imageSpec == nil {
+		return nil
+	}
+	if extraMount.GetHostPath() != "" {
+		return fmt.Errorf("hostpath must be empty while mount image: %+v", extraMount)
+	}
+	if !extraMount.GetReadonly() {
+		return fmt.Errorf("readonly must be true while mount image: %+v", extraMount)
+	}
+
+	ref := imageSpec.GetImage()
+	if ref == "" {
+		return fmt.Errorf("image not specified in: %+v", imageSpec)
+	}
+	image, err := c.LocalResolve(ref)
+	if err != nil {
+		return fmt.Errorf("failed to resolve image %q: %w", ref, err)
+	}
+	containerdImage, err := c.toContainerdImage(ctx, image)
+	if err != nil {
+		return fmt.Errorf("failed to get image from containerd %q: %w", image.ID, err)
+	}
+
+	// This is a digest of the manifest
+	imageID := containerdImage.Target().Digest.Encoded()
+
+	target := c.getImageVolumeHostPath(sandboxID, imageID)
+
+	// Already mounted in another container on the same pod
+	if stat, err := os.Stat(target); err == nil && stat.IsDir() {
+		extraMount.HostPath = target
+		return nil
+	}
+
+	ctx, done, err := c.client.WithLease(ctx,
+		leases.WithID(target),
+		leases.WithLabel(defaults.DefaultSnapshotterNSLabel, snapshotter),
+		leases.WithLabel(labelGCSnapRef+snapshotter, target),
+	)
+	if err != nil && !errdefs.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create lease: %w", err)
+	}
+	defer func() {
+		if retErr != nil {
+			_ = done(ctx) // call lease done func.
+		}
+	}()
+
+	img, err := c.client.ImageService().Get(ctx, ref)
+	if err != nil {
+		return fmt.Errorf("failed to get image volume ref %q: %w", ref, err)
+	}
+
+	i := containerd.NewImageWithPlatform(c.client, img, platforms.Only(platform))
+	if err := i.Unpack(ctx, snapshotter); err != nil {
+		return fmt.Errorf("failed to unpack image volume: %w", err)
+	}
+
+	diffIDs, err := i.RootFS(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get diff IDs for image volume %q: %w", ref, err)
+	}
+	chainID := identity.ChainID(diffIDs).String()
+
+	s := c.client.SnapshotService(snapshotter)
+	mounts, err := s.Prepare(ctx, target, chainID)
```

Comment:
Are these intentionally writable or is there a flag to make this readonly?

If this is supposed to be writable, where is that space accounted for?

Reply:
This is intentional, it requires the SELinux to be set for it, which is currently not possible with a readonly.

Reply:
I think it would work to set the mount path to writable, then set SELinux, then set it to readonly.

However, I think it's a big change and not easily accepted. I can explore its feasibility later if necessary.

Reply:
@dmcgowan is it acceptable to merge this for now and address with a follow up?

Reply:
agree to get this as first iteration. we can handle all the questions in the follow-up